plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,842 @@
+                "routines!"
+            )
+
+        path = self.output_directory / f"{self._particle_tracker.iteration_number}.hdf5"
+
+        with h5py.File(path, "w") as output_file:
+            output_file["x"] = self._particle_tracker.x
+            output_file["v"] = self._particle_tracker.v
+
+    def save_to_memory(self):
+        """Append simulation positions and velocities to save routine object."""
+        self.x_all.append(np.copy(self._particle_tracker.x))
+        self.v_all.append(np.copy(self._particle_tracker.v))
+
+    def post_push_hook(self, force_save=False):
+        """Function called after a push step.
+
+        This function is responsible for handling two steps of save routine, namely:
+            - Deciding whether or not to save on the current time step
+            - How the simulation data is saved (i.e. to disk or memory)
+
+        """
+        if self.save_now or force_save:
+            self.save()
+
+
+class AbstractIntervalSaveRoutine(AbstractSaveRoutine, ABC):
+    """Abstract class describing a save routine that saves every given interval."""
+
+    def __init__(self, interval: u.Quantity, **kwargs):
+        super().__init__(**kwargs)
+        self.save_interval = interval.to(u.s).value
+        self.time_of_last_save = 0
+
+    @property
+    def require_uniform_dt(self):
+        """Save output only makes sense for uniform time steps,
+        therefore this routine requires a uniform time step.
+        """
+        return True
+
+    @property
+    def save_now(self):
+        """Save at every interval given in instantiation."""
+        saves_per_step = np.floor(self.save_interval / self._particle_tracker.dt)
+
+        # If the user is requesting multiple saves per step then raise a ValueError
+        if saves_per_step == 0:
+            raise ValueError(
+                "You must specify a time step smaller than the save interval!"
+            )
+
+        if self._particle_tracker.time - self.time_of_last_save >= self.save_interval:
+            self.time_of_last_save = self._particle_tracker.time
+
+            return True
+        else:
+            return False
+
+
+class DiskIntervalSaveRoutine(AbstractIntervalSaveRoutine):
+    """Save routine corresponding to saving a hdf5 file every given interval."""
+
+    def __init__(self, interval: u.Quantity, output_directory: Path):
+        super().__init__(interval, output_directory=output_directory)
+
+    def save(self):
+        """Save the state of the particle tracker to disk."""
+        self.save_to_disk()
+
+
+class MemoryIntervalSaveRoutine(AbstractIntervalSaveRoutine):
+    """Save the state of the tracker every given interval."""
+
+    def __init__(self, interval: u.Quantity):
+        super().__init__(interval)
+
+    def save(self):
+        """Save the state of the particle tracker to memory."""
+        self.save_to_memory()
+
+
+class ParticleTracker:
```

Comment:
This whole class needs a docstring.

Can you add notes to the docstring for this class defining what the NaNs mean (for velocity and position). If both are NaN it means the particle was removed: what if it is stopped?

Reply:
I don't think we're using the stopping or removing mechanisms yet. Should I remove them from this pull request and put them somewhere for a later PR?